web/src/OverviewLogPane.tsx

```diff
@@ -320,27 +372,150 @@ export class OverviewLogComponent extends Component<OverviewLogComponentProps> {
       }
       return true
     })
+
     this.logCheckpoint = patch.checkpoint
+    lines.forEach((line) => this.lineHashList.append(line))
+
+    if (startCheckpoint) {
+      // If this is an incremental render, put the lines in the forward buffer.
+      lines.forEach((line) => {
+        this.forwardBuffer.push(line)
+      })
+    } else {
+      // If this is the first render, put the lines in the backward buffer, so
+      // that the last lines get rendered first.
+      lines.forEach((line) => {
+        this.backwardBuffer.push(line)
+      })
+    }
+
+    if (this.renderBufferRafId) return
+    this.renderBufferRafId = this.props.raf.requestAnimationFrame(
+      this.renderBuffer
+    )
+  }
+
+  // We have two render buffers:
+  // - a buffer of newer logs that we haven't rendered yet.
+  // - a buffer of older logs that we haven't rendered yet.
+  // First, process the newer logs.
+  // If we're out of new logs to render, go back through the old logs.
+  renderBuffer() {
+    this.renderBufferRafId = 0
+
+    let root = this.rootRef.current
+    let cursor = this.cursorRef.current
+    if (!root || !cursor) {
+      return
+    }
 
-    let lastManifestName = ""
-    for (let i = 0; i < lines.length; i++) {
-      let line = lines[i]
-      let isContextChange = i > 0 && line.manifestName !== lastManifestName
-      let lineEl = newLineEl(line, isContextChange, showManifestName)
-      let existingLineEl = this.elByStoredLineIndex[line.storedLineIndex]
-      if (existingLineEl) {
-        root.replaceChild(lineEl, existingLineEl)
-      } else {
-        root.insertBefore(lineEl, cursor)
+    // If there are no lines in either buffer, we're done.
+    if (!this.backwardBuffer.length && !this.forwardBuffer.length) {
+      return
+    }
+
+    // Render the lines in the forward buffer first.
+    let forwardLines = this.forwardBuffer.slice(0, renderWindow)
+    this.forwardBuffer = this.forwardBuffer.slice(renderWindow)
+    for (let i = 0; i < forwardLines.length; i++) {
+      let line = forwardLines[i]
+      let rendered = this.renderLineHelper(line)
+      if (!rendered) {
+        // This must be updating the progress of something that hasn't been
+        // rendered yet.
```

Comment:
so we're dropping this line in the backward buffer, despite it still being a "new" log line that needs rendering, on the assumption that by the time we get around to processing the backward buffer, we may have rendered the line that this line updates?

Reply:
ya